README.md
# @waku-objects/luminance

[![Tests](https://github.com/logos-innovation-lab/luminance/actions/workflows/test.yml/badge.svg)](https://github.com/logos-innovation-lab/luminance/actions/workflows/test.yml)
[![](https://img.shields.io/badge/made%20by-Logos%20Innovation%20Lab-blue.svg?style=flat-square)](https://github.com/logos-innovation-lab)
[![standard-readme compliant](https://img.shields.io/badge/standard--readme-OK-brightgreen.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/feross/standard)
![](https://img.shields.io/badge/pnpm-%3E%3D7.0.0-orange.svg?style=flat-square)
![](https://img.shields.io/badge/Node.js-%3E%3D16.0.0-orange.svg?style=flat-square)
![](https://img.shields.io/badge/runs%20in-browser%20%7C%20node%20%7C%20webworker%20%7C%20electron-orange)

> Library for calculating color luminance and getting corresponding luminance colors

**Warning: This project is in beta state. There might (and most probably will) be changes in the future to its API and working. Also, no guarantees can be made about its stability, efficiency, and security at this stage.**

## Table of Contents

- [Install](#install)
- [Usage](#usage)
- [License](#license)

## Install

```sh
pnpm install @waku-objects/luminance --save
```

## Usage

```typescript
import { calculateLuminance, getClosestColor } from '@waku-objects/luminance'

// Calculate luminance of a color (luminance is between 0 and 1)
calculateLuminance('#acacac') // supports hex
calculateLuminance({ r: 200, g: 200, b: 200 }) // and RGB

// Finds a color with the same hue that would fullfil target luminance
getClosestColor('#ff0000', 0.5) // returns #ffa2a2
getClosestColor({ r: 255, g: 0, b: 9 }, 0.5) // returns #ffa2a2
```

List of all functions and interfaces:

```typescript
interface RGB {
	r: number
	g: number
	b: number
}
interface HSL {
	h: number
	s: number
	l: number
}
export interface HUE {
	p: number
	q: number
	t: number
}

// Utility conversion functions
function hexToRgb(hex: string): RGB
function rgbToHex(rgb: RGB): string
function hslToRgb(hsl: HSL): RGB
function rgbToHsl(rgb: RGB): HSL

// Calculate luminance
function calculateLuminance(color: RGB | string): number

// Find color with same hue that would be closest to a targetLuminance (uses bisection algorithm with at most 100 iterations)
function getClosestColor(
	sourceColor: string,
	targetLuminance: number,
	targetPrecision?: number,
	maxSteps?: number
): string

function getClosestColorBisection(
	sourceColor: string,
	targetLuminance: number,
	targetPrecision?: number,
	maxSteps?: number
): string

// Implementation with newton itterative method, better results than bisection for more than 100 iterations
function getClosestColorNewton(
	sourceColor: string,
	targetLuminance: number,
	targetPrecision?: number,
	maxSteps?: number
): string
```

## License

[MIT](./LICENSE)


CHANGELOG.md
# Changelog

## [2.0.1](https://github.com/logos-innovation-lab/luminance/compare/v2.0.0...v2.0.1) (2023-07-21)


### Bug Fixes

* indicate this is ES Module and not transpiled to Common JS ([#17](https://github.com/logos-innovation-lab/luminance/issues/17)) ([40484a4](https://github.com/logos-innovation-lab/luminance/commit/40484a404f44ec12c8083ab11df0b3b78a5d6b8d))

## [2.0.0](https://github.com/logos-innovation-lab/luminance/compare/v1.2.0...v2.0.0) (2023-07-21)


### ⚠ BREAKING CHANGES

* typescript should be dev depency ([#13](https://github.com/logos-innovation-lab/luminance/issues/13))

### Bug Fixes

* expose types ([#16](https://github.com/logos-innovation-lab/luminance/issues/16)) ([9b155b3](https://github.com/logos-innovation-lab/luminance/commit/9b155b3d41ee5a87c9476880a3f021dfa5302b64))
* typescript should be dev depency ([#13](https://github.com/logos-innovation-lab/luminance/issues/13)) ([738ca99](https://github.com/logos-innovation-lab/luminance/commit/738ca99db531e0a968c4c6ce968c8aa950c84c16))

## [1.2.0](https://github.com/logos-innovation-lab/luminance/compare/v1.1.0...v1.2.0) (2023-07-21)


### Features

* add support for rgb object to getClosestColor ([#10](https://github.com/logos-innovation-lab/luminance/issues/10)) ([9937137](https://github.com/logos-innovation-lab/luminance/commit/9937137aac9ee38747536e26620d63caf5bb898d))

## [1.1.0](https://github.com/logos-innovation-lab/luminance/compare/v1.0.0...v1.1.0) (2023-07-20)


### Features

* extract default precision and steps ([#5](https://github.com/logos-innovation-lab/luminance/issues/5)) ([a2bdd79](https://github.com/logos-innovation-lab/luminance/commit/a2bdd79ff47672ba4cd8760ccd3f2c34c8d543ac))

## 1.0.0 (2023-07-20)


### Features

* find closest color to luminance and calculate luminance ([#1](https://github.com/logos-innovation-lab/luminance/issues/1)) ([d1ded48](https://github.com/logos-innovation-lab/luminance/commit/d1ded48119e97ede1b4cdf39d91f0ef98eb9c125))


test/index.spec.ts
import { describe, it, expect } from 'vitest'
import {
	hexToRgb,
	rgbToHex,
	hslToRgb,
	calculateLuminance,
	getClosestColor,
	getClosestColorNewton,
	type RGB,
	type HSL,
	rgbToHsl
} from '../src/index'

function formatRgb(rgb: RGB) {
	return `{r=${rgb.r.toString().padStart(3, ' ')}, g=${rgb.g.toString().padStart(3, ' ')}, b=${rgb.b
		.toString()
		.padStart(3, ' ')}}`
}
function formatHsl(hsl: HSL) {
	return `{h=${hsl.h.toString().padStart(3, ' ')}, s=${hsl.s.toString().padStart(3, ' ')}, l=${hsl.l
		.toString()
		.padStart(3, ' ')}}`
}

describe('hexToRgb', () => {
	const testValues = [
		{ hex: '#000000', rgb: { r: 0, g: 0, b: 0 } },
		{ hex: '#ffffff', rgb: { r: 255, g: 255, b: 255 } },
		{ hex: '#ff0000', rgb: { r: 255, g: 0, b: 0 } },
		{ hex: '#00ff00', rgb: { r: 0, g: 255, b: 0 } },
		{ hex: '#0000ff', rgb: { r: 0, g: 0, b: 255 } }
	]

	testValues.forEach(({ hex, rgb }) => {
		it(`with hex=${hex} should return ${formatRgb(rgb)}`, () => {
			const res = hexToRgb(hex)
			expect(res.r).toEqual(rgb.r)
			expect(res.g).toEqual(rgb.g)
			expect(res.b).toEqual(rgb.b)
		})
	})
})

describe('rgbToHex', () => {
	const testValues = [
		{ rgb: { r: 0, g: 0, b: 0 }, hex: '#000000' },
		{ rgb: { r: 255, g: 255, b: 255 }, hex: '#ffffff' },
		{ rgb: { r: 255, g: 0, b: 0 }, hex: '#ff0000' },
		{ rgb: { r: 0, g: 255, b: 0 }, hex: '#00ff00' },
		{ rgb: { r: 0, g: 0, b: 255 }, hex: '#0000ff' }
	]

	testValues.forEach(({ rgb, hex }) => {
		it(`with rgb = ${formatRgb(rgb)} should return ${hex}`, () => {
			const res = rgbToHex(rgb)
			expect(res).toEqual(hex)
		})
	})
})

describe('hslToRgb', () => {
	const testValues = [
		{ hsl: { h: 0, s: 0, l: 0 }, rgb: { r: 0, g: 0, b: 0 } },
		{ hsl: { h: 0, s: 0, l: 1 }, rgb: { r: 255, g: 255, b: 255 } },
		{ rgb: { r: 255, g: 0, b: 0 }, hsl: { h: 0, s: 1, l: 0.5 } },
		{ rgb: { r: 0, g: 255, b: 0 }, hsl: { h: 0.3333333333333333, s: 1, l: 0.5 } },
		{ rgb: { r: 0, g: 0, b: 255 }, hsl: { h: 0.6666666666666666, s: 1, l: 0.5 } }
	]

	testValues.forEach(({ hsl, rgb }) => {
		it(`with hsl=${formatHsl(hsl)} should return ${formatRgb(rgb)}`, () => {
			const res = hslToRgb(hsl)
			expect(res.r).toEqual(rgb.r)
			expect(res.g).toEqual(rgb.g)
			expect(res.b).toEqual(rgb.b)
		})
	})
})

describe('rgbToHsl', () => {
	const testValues = [
		{ rgb: { r: 0, g: 0, b: 0 }, hsl: { h: 0, s: 0, l: 0 } },
		{ rgb: { r: 255, g: 255, b: 255 }, hsl: { h: 0, s: 0, l: 1 } },
		{ rgb: { r: 255, g: 0, b: 0 }, hsl: { h: 0, s: 1, l: 0.5 } },
		{ rgb: { r: 0, g: 255, b: 0 }, hsl: { h: 0.3333333333333333, s: 1, l: 0.5 } },
		{ rgb: { r: 0, g: 0, b: 255 }, hsl: { h: 0.6666666666666666, s: 1, l: 0.5 } }
	]

	testValues.forEach(({ rgb, hsl }) => {
		it(`with rgb = ${formatRgb(rgb)} should return ${formatHsl(hsl)}`, () => {
			const res = rgbToHsl(rgb)
			expect(res.h).toEqual(hsl.h)
			expect(res.s).toEqual(hsl.s)
			expect(res.l).toEqual(hsl.l)
		})
	})
})

describe('calculateLuminance', () => {
	const testValues = [
		{ rgb: { r: 255, g: 255, b: 255 }, luminance: 1 },
		{ rgb: { r: 0, g: 0, b: 0 }, luminance: 0 },
		{ rgb: { r: 255, g: 0, b: 0 }, luminance: 0.2126 },
		{ rgb: { r: 0, g: 255, b: 0 }, luminance: 0.7152 },
		{ rgb: { r: 0, g: 0, b: 255 }, luminance: 0.0722 }
	]

	testValues.forEach(({ rgb, luminance }) => {
		it(`with rgb = ${formatRgb(rgb)} should return ${luminance}`, () => {
			const res = calculateLuminance(rgb)
			expect(res).toEqual(luminance)
		})
	})
})

describe('getClosestColor', () => {
	const testValues = [
		{ hex: '#000000', luminance: 0, hexOut: '#000000' },
		{ hex: '#000000', luminance: 1, hexOut: '#ffffff' },
		{ hex: '#000000', luminance: 0.5, hexOut: '#bcbcbc' },
		{ hex: '#ff0000', luminance: 0.5, hexOut: '#ffa2a2' },
		{ hex: '#00ff00', luminance: 0.5, hexOut: '#00d900' },
		{ hex: '#0000ff', luminance: 0.5, hexOut: '#b5b5ff' }
	]

	testValues.forEach(({ hex, luminance, hexOut }) => {
		it(`with hex=${hex} and luminance target ${luminance} should return ${hexOut}`, () => {
			const res = getClosestColor(hex, luminance)
			expect(res).toEqual(hexOut)
		})
	})
})

describe('getClosestColorNewton', () => {
	const testValues = [
		{ hex: '#000000', luminance: 0, hexOut: '#000000' },
		{ hex: '#000000', luminance: 1, hexOut: '#ffffff' },
		{ hex: '#000000', luminance: 0.5, hexOut: '#bbbbbb' },
		{ hex: '#ff0000', luminance: 0.5, hexOut: '#ffa2a2' },
		{ hex: '#00ff00', luminance: 0.5, hexOut: '#00d900' },
		{ hex: '#0000ff', luminance: 0.5, hexOut: '#b4b4ff' }
	]

	testValues.forEach(({ hex, luminance, hexOut }) => {
		it(`with hex=${hex} and luminance target ${luminance} should return ${hexOut}`, () => {
			const res = getClosestColorNewton(hex, luminance)
			expect(res).toEqual(hexOut)
		})
	})
})


src/index.ts
export interface RGB {
	r: number
	g: number
	b: number
}

export interface HSL {
	h: number
	s: number
	l: number
}

export interface HUE {
	p: number
	q: number
	t: number
}

const DEFAULT_MAX_STEPS = 100
const DEFAULT_TARGET_PRECISION = 0.01

export function hexToRgb(hex: string): RGB {
	const bigint = parseInt(hex.substring(1), 16)
	const r = (bigint >> 16) & 255
	const g = (bigint >> 8) & 255
	const b = bigint & 255

	return { r, g, b }
}

export function rgbToHex({ r, g, b }: RGB) {
	return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b
		.toString(16)
		.padStart(2, '0')}`
}

// Converts an HSL color value to RGB.
// h, s, and l are contained in the set [0, 1] and
// returns r, g, and b in the set [0, 255].
export function hslToRgb({ h, s, l }: HSL): RGB {
	let r, g, b

	if (s == 0) {
		r = g = b = l // achromatic
	} else {
		const q = l < 0.5 ? l * (1 + s) : l + s - l * s
		const p = 2 * l - q
		r = hue2rgb({ p, q, t: h + 1 / 3 })
		g = hue2rgb({ p, q, t: h })
		b = hue2rgb({ p, q, t: h - 1 / 3 })
	}

	return { r: Math.round(r * 255), g: Math.round(g * 255), b: Math.round(b * 255) }
}

// Converts an RGB color value to HSL.
// r, g, and b are contained in the set [0, 255] and
// returns h, s, and l in the set [0, 1].
export function rgbToHsl({ r, g, b }: RGB): HSL {
	r /= 255
	g /= 255
	b /= 255
	const max = Math.max(r, g, b)
	const min = Math.min(r, g, b)
	const l = (max + min) / 2
	let h,
		s = l

	if (max == min) {
		h = s = 0 // achromatic
	} else {
		const d = max - min
		s = l > 0.5 ? d / (2 - max - min) : d / (max + min)
		switch (max) {
			case r:
				h = (g - b) / d + (g < b ? 6 : 0)
				break
			case g:
				h = (b - r) / d + 2
				break
			default: // case b:
				h = (r - g) / d + 4
				break
		}
		h /= 6
	}

	return { h, s, l }
}

function hue2rgb({ p, q, t }: HUE) {
	if (t < 0) t += 1
	if (t > 1) t -= 1
	if (t < 1 / 6) return p + (q - p) * 6 * t
	if (t < 1 / 2) return q
	if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6
	return p
}

function gammaCorrect(color: number) {
	return color <= 0.03928 ? color / 12.92 : ((color + 0.055) / 1.055) ** 2.4
}

export function calculateLuminance(sourceColor: RGB | string) {
	if (typeof sourceColor === 'string') sourceColor = hexToRgb(sourceColor)
	let { r, g, b } = sourceColor

	// Normalized to values between 0 and 1
	r /= 255
	g /= 255
	b /= 255

	// apply gamma correction
	r = gammaCorrect(r)
	g = gammaCorrect(g)
	b = gammaCorrect(b)

	// calculate luminance
	return 0.2126 * r + 0.7152 * g + 0.0722 * b
}

export function getClosestColor(
	sourceColor: string | RGB,
	targetLuminance: number,
	targetPrecision = DEFAULT_TARGET_PRECISION,
	maxSteps = DEFAULT_MAX_STEPS
) {
	return getClosestColorBisection(sourceColor, targetLuminance, targetPrecision, maxSteps)
}

export function getClosestColorNewton(
	sourceColor: string | RGB,
	targetLuminance: number,
	targetPrecision = DEFAULT_TARGET_PRECISION,
	maxSteps = DEFAULT_MAX_STEPS
) {
	// 	// Just a shortcut for all colors these are 0 and 100 luminance
	if (targetLuminance <= 0) return '#000000'
	if (targetLuminance >= 1) return '#ffffff'

	const hsl = rgbToHsl(typeof sourceColor === 'string' ? hexToRgb(sourceColor) : sourceColor)
	const damping = 15 // Damping factor to avoid instability

	for (let steps = 0; steps < maxSteps; ++steps) {
		// Limit iterations to prevent infinite loop
		const rgb = hslToRgb(hsl)
		const currentLuminance = calculateLuminance(rgb)
		const error = currentLuminance - targetLuminance

		if (Math.abs(error) / targetLuminance <= targetPrecision) {
			return rgbToHex(rgb) // Found suitable luminance
		}

		// If the color is white and the target luminance is less than the current, start decreasing the brightness
		if (rgb.r === 255 && rgb.g === 255 && rgb.b === 255 && targetLuminance < currentLuminance) {
			hsl.l -= 0.01 // decrement brightness
			continue
		}

		// Newton-Raphson update
		const tmpRgb = hslToRgb({ ...hsl, l: hsl.l + 0.01 })
		const derivative = (calculateLuminance(tmpRgb) - currentLuminance) / 0.01

		if (error) hsl.l -= error / derivative / damping

		// Clamp brightness to [0,1]
		if (hsl.l < 0) hsl.l = 0
		if (hsl.l > 1) hsl.l = 1
	}

	return rgbToHex(hslToRgb(hsl)) // Suitable luminance wasn't found in time, return
}

export function getClosestColorBisection(
	sourceColor: string | RGB,
	targetLuminance: number,
	targetPrecision = DEFAULT_TARGET_PRECISION,
	maxSteps = DEFAULT_MAX_STEPS
) {
	// Just a shortcut for all colors these are 0 and 100 luminance
	if (targetLuminance <= 0) return '#000000'
	if (targetLuminance >= 1) return '#ffffff'

	const hsl = rgbToHsl(typeof sourceColor === 'string' ? hexToRgb(sourceColor) : sourceColor)

	let low = 0,
		high = 1,
		steps = 0
	let lowLum = 0,
		highLum = 1
	for (
		steps = 0;
		(highLum - lowLum) / targetLuminance > targetPrecision && steps < maxSteps;
		steps += 1
	) {
		// Stop when interval size is within target precision
		hsl.l = (low + high) / 2 // Take the midpoint as the new guess
		const currentLuminance = calculateLuminance(hslToRgb(hsl))

		if (currentLuminance > targetLuminance) {
			high = hsl.l
			highLum = currentLuminance
		} else {
			low = hsl.l
			lowLum = currentLuminance
		}
	}

	return rgbToHex(hslToRgb(hsl)) // Return the hex color for the best luminance match
}
